### PR TITLE
Implement cyclone mortality scenarios

### LINKED ISSUE #531: Use cyclone mortality model and filter to implement cyclone mortality in ADRIAmod
Part of #516 .

### PR DESCRIPTION
**Pull request by Zapiano**
- [x] Load cyclone mortality datacube (using a new data package version) into ADRIA 
- [x] Use it to add coral mortality scenarios caused by cyclones
- [x] Adapt tests to use new Datapackage version

Closes #531 .

### CODE REVIEW

**Comment by Zapiano** [comment]
@ConnectedSystems @Rosejoycrocker 

Note that this is still a draft. The only thing I did so far was load the cyclone datacube.

**Comment by Zapiano** [comment]
@ConnectedSystems  this is virtually ready but I can't fix the tests as they are broken at the main branch. As you know, I also can't show the results yet (and be 100% sure they make sense) as our scenarios visualization function is also broken. I'm waiting for all of that to be fixed before I finish this.

**Comment by ConnectedSystems** [comment]
Depends on #589 being merged first.

**Comment by Zapiano** [comment]
@ConnectedSystems and @Rosejoycrocker  this one is (finally) ready for review. Here is an example of result with 1024 scenarios:

image

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]
A few changes necessary.

How do you feel about merging PR #597 into this PR?

Next set of work could be to rename `Example_domain` to `Test_domain`, move it, and include a bunch of checks to that raise error when a malformed Domain is provided by the user?

**Comment by ConnectedSystems** [comment]
I've added a test that checks the location order but it currently fails as the example geospatial data holds a different set of location names compared to those specified in the netcdf:

These are the ones in the example geospatial data:

```
 "Briggs_BR_C_1"
 "Briggs_BR_OF_1"
 "Briggs_BR_OF_2"
 "Briggs_BR_SS_1"
 "Briggs_BR_S_1"
 "Briggs_BR_S_2"
 "Briggs_BR_S_3"
 "Elford_ER_C_1"
 "Elford_ER_C_10"
 "Elford_ER_C_11"
 "Elford_ER_C_12"
 "Elford_ER_C_2"
 "Elford_ER_C_3"
 "Elford_ER_C_4"
 "Elford_ER_C_5"
 "Elford_ER_C_6"
 "Elford_ER_C_7"
 "Elford_ER_C_8"
 "Elford_ER_C_9"
 "Elford_ER_OF_1"
 "Elford_ER_OF_10"
 "Elford_ER_OF_11"
 "Elford_ER_OF_12"
 "Elford_ER_OF_13"
 "Elford_ER_OF_14"
 "Elford_ER_OF_15"
 "Elford_ER_OF_16"
 "Elford_ER_OF_17"
 "Elford_ER_OF_18"
 "Elford_ER_OF_19"
 "Elford_ER_OF_2"
 "Elford_ER_OF_3"
```

**Comment by Zapiano** [comment]
> A few changes necessary.
> 
> How do you feel about merging PR #597 into this PR?
> 
> Next set of work could be to rename `Example_domain` to `Test_domain`, move it, and include a bunch of checks to that raise error when a malformed Domain is provided by the user?
> 
> I'm also adding tests to check that handling of cyclone data is correct.

Honestly, I don't think we should or need to merge #597 into this one. They are not directly related, and it's easier/better to do small PRs when possible. I don't see what's the advantage, but I do see some risks.

File: src/ADRIA.jl [hunk 1]
```diff
@@ -101,7 +102,7 @@ export ReefModDomain
 # export dims, ndims
 
 # List out compatible domain datapackages
-const COMPAT_DPKG = ["0.3.1"]
+const COMPAT_DPKG = ["0.3.2-rc"]
```
[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/ADRIA.jl:105]: Oh actually, I think we should bump to "0.4.0" / "0.4.0-rc"

**Comment by ConnectedSystems** [comment]
> > A few changes necessary.
> > How do you feel about merging PR #597 into this PR?
> > Next set of work could be to rename `Example_domain` to `Test_domain`, move it, and include a bunch of checks to that raise error when a malformed Domain is provided by the user?
> > I'm also adding tests to check that handling of cyclone data is correct.
> 
> Honestly, I don't think we should or need to merge #597 into this one. They are not directly related, and it's easier/better to do small PRs when possible. I don't see what's the advantage, but I do see some risks.

Well sure, it is two small PRs, it's just one goes into the other. But happy to target main with #597

File: src/ADRIA.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (Zapiano) [reply src/ADRIA.jl]: About being 0.4.0: I agree, I can update that.

About support 0.3.1: sure. I can update the loading method to handle that.

**Comment by Zapiano** [comment]
@ConnectedSystems ready for another review :)

File: src/io/inputs.jl [hunk 2]
```diff
@@ -225,3 +227,61 @@ function load_env_data(data_fn::String, attr::String, site_data::DataFrame)::Nam
 
     return data
 end
+
+"""
+    load_cyclone_mortality(data_fn::String)::NamedDimsArray
```
Comment (ConnectedSystems) [inline src/io/inputs.jl:232]:
```suggestion
    load_cyclone_mortality(data_fn::String)::NamedDimsArray
    load_cyclone_mortality(timeframe::Vector{Int64}, site_data::DataFrame)::NamedDimsArray
```

File: src/scenario.jl [hunk 3]
```diff
@@ -776,3 +783,21 @@ function run_model(domain::Domain, param_set::NamedDimsArray, corals::DataFrame,
     site_ranks[site_ranks.==0.0] .= n_locs + 1
     return (raw=C_cover, seed_log=Yseed, fog_log=Yfog, shade_log=Yshade, site_ranks=site_ranks, bleaching_mortality=bleaching_mort, coral_dhw_log=collated_dhw_tol_log)
 end
+
+function _cyclone_mortality(coral_cover, coral_params, cyclone_mortality)::Nothing
```
Comment (ConnectedSystems) [inline src/scenario.jl:787]: I don't think this is a private function as such.

The entire `run_model()` needs a clean up, but the idea would be to separate out each event into their own function and run in order of their envisioned sequence 

For example:

```
shade_locations!()
# ... etc...

bleaching_mortality!()
cyclone_mortality!()
# ... etc...

# run ODE
# ...
```

Comment (ConnectedSystems) [reply src/scenario.jl:787]: This function modifies in place so by convention it should have the exclamation mark

```suggestion
function _cyclone_mortality!(coral_cover, coral_params, cyclone_mortality)::Nothing
```

**Review by ConnectedSystems — APPROVED** [review]
One minor change after which I'll merge.

Otherwise all tests pass, including the figure generation script.

**Comment by ConnectedSystems** [comment]
Done.

**Pull request merged**